package/piksi_system_daemon/src/async-child.c

```diff
@@ -0,0 +1,163 @@
+ */
+static int raw_fgets(char *str, size_t len, int fd)
+{
+  size_t i;
+  len--;
+  for (i = 0; i < len; i++) {
+    int r = read(fd, &str[i], 1);
+    if ((r < 0) && (errno != EAGAIN))
+      return r;
+    if ((r <= 0) || (str[i] == '\n'))
+      break;
+  }
+  str[i++] = 0;
+  return i > 1 ? i : 0;
+}
+
+static int async_output_cb(zloop_t *loop, zmq_pollitem_t *item, void *arg)
+{
+  struct async_child_ctx *ctx = arg;
+  char buf[128];
+
+  if (raw_fgets(buf, sizeof(buf), item->fd) > 0) {
+    if (ctx->output_callback)
+      ctx->output_callback(buf, ctx->external_context);
+  }
+
+  return 0;
+}
+
+static int async_cleanup_dead_child(zloop_t *loop, zmq_pollitem_t *item, void *arg)
+{
+  struct async_child_ctx *ctx;
+  read(item->fd, &ctx, sizeof(ctx));
+
+  /* Child is dead, notify observer and clean up */
+  ctx->exit_callback(ctx->exit_status, ctx->external_context);
+
+  if (ctx == async_children) {
+    async_children = ctx->next;
+  } else {
+    for (struct async_child_ctx *c = async_children; c; c = c->next) {
+      if (c->next == ctx) {
+        c->next = ctx->next;
+        break;
+      }
+    }
+  }
+  free(ctx);
+
+  return 0;
+}
+
+int async_spawn(zloop_t *loop, char **argv,
+                void (*output_callback)(const char *buf, void *ctx),
+                void (*exit_callback)(int status, void *ctx),
+                void *external_context)
+{
+  if (!wakepipe[0]) {
+    /* Set up the wake up pipe if not done already. */
+    pipe(wakepipe);
+    wakeitem.fd = wakepipe[0];
+    wakeitem.events = ZMQ_POLLIN;
+    zloop_poller(loop, &wakeitem, async_cleanup_dead_child, NULL);
+  }
+
+  struct async_child_ctx *ctx = calloc(1, sizeof(*ctx));
+  ctx->loop = loop;
+  ctx->output_callback = output_callback;
+  ctx->exit_callback = exit_callback;
+  ctx->external_context = external_context;
+
+  /* Open pipe to receive child's output */
+  int pipefd[2];
+  if (pipe(pipefd)) {
+    perror("async_spawn pipe");
+    free(ctx);
+    return -1;
+  }
+
+  ctx->pid = fork();
+  if (ctx->pid == 0) {
+    /* Code executed by new child */
+    close(STDIN_FILENO);
+    dup2(pipefd[1], STDOUT_FILENO);
+    execvp(argv[0], argv);
+    perror("async_spawn exec");
+    exit(1);
+  }
+
+  close(pipefd[1]);
+
+  if (ctx->pid == -1) {
+    /* Failed to fork */
+    perror("async_spawn fork");
+    close(pipefd[0]);
+    free(ctx);
+    return -1;
+  }
+
+  ctx->pollitem.fd = pipefd[0];
+  ctx->pollitem.events = ZMQ_POLLIN | ZMQ_POLLERR;
+  int arg = O_NONBLOCK;
+  fcntl(pipefd[0], F_SETFL, &arg);
+  zloop_poller(loop, &ctx->pollitem, async_output_cb, ctx);
+
+  /* Add to child list */
+  ctx->next = async_children;
+  async_children = ctx;
+}
+
+void async_child_waitpid_handler(pid_t pid, int status)
+{
+  /* Called by SIGCHLD handler with status from waitpid.
+   * We save the status here, and will deal with it from the zmq loop by
+   * writing to the wakeup pipe.
+   */
+  for (struct async_child_ctx *c = async_children; c; c = c->next) {
```

Comment:
Since this function is executed from the `SIGCHLD` handler, you should mask `SIGCHLD` while modifying the `async_chld_ctx` list from other parts of the code. E.g. https://github.com/swift-nav/piksi_buildroot/blob/master/package/piksi_system_daemon/src/ports.c#L151